src/MOI/attributes.jl

```diff
@@ -0,0 +1,259 @@
+        push!(out, (MOI.ScalarAffineFunction{Float64}, MOI.GreaterThan{Float64}))
+    end
+    if length(constraint_storage(m, MOI.ScalarAffineFunction{Float64}, MOI.EqualTo{Float64})) > 0
+        push!(out, (MOI.ScalarAffineFunction{Float64}, MOI.EqualTo{Float64}))
+    end
+
+    # TODO variablewise
+
+    return out
+end
+
+# """
+#     ObjectiveFunction()
+# An `AbstractFunction` instance which represents the objective function.
+# It is guaranteed to be equivalent but not necessarily identical to the function provided by the user.
+# """
+# struct ObjectiveFunction <: AbstractSolverInstanceAttribute end
+
+
+## Termination status
+# """
+#     TerminationStatus()
+# A `TerminationStatusCode` explaining why the solver stopped.
+# """
+# struct TerminationStatus <: AbstractSolverInstanceAttribute end
+MOI.cangetattribute(m::XpressSolverInstance, ::MOI.TerminationStatus) = true
+function MOI.getattribute(m::XpressSolverInstance, ::MOI.TerminationStatus)
+    if !is_mip(m.inner)
+        return terminationcode_lp[get_lp_status(m.inner)]
+    else
+        return terminationcode_mip[get_mip_status(m.inner)]
+    end
+end
+
+# see STOPSTATUS
+const terminationcode_lp = Dict(
+    :unstarted        => MOI.OtherError, # TODO
+    :optimal          => MOI.Success,
+    :infeasible       => MOI.InfeasibleNoResult, # TODO improve
+    :cutoff           => MOI.OtherError, # TODO
+    :unfinished       => MOI.OtherError, # TODO
+    :unbounded        => MOI.UnboundedNoResult, # TODO improve
+    :cutoff_in_dual   => MOI.OtherError, # TODO
+    :unsolved         => MOI.OtherError, # TODO
+    :nonconvex        => MOI.InvalidSolverInstance # TODO
+)
+const terminationcode_mip = Dict(
+    :mip_not_loaded     => MOI.OtherError, # TODO
+    :mip_lp_not_optimal => MOI.OtherError, # TODO improve
+    :mip_lp_optimal     => MOI.OtherError, # TODO improve
+    :mip_no_sol_found   => MOI.InfeasibleNoResult, # TODO improve
+    :mip_suboptimal     => MOI.AlmostSuccess, #TODO
+    :mip_infeasible     => MOI.InfeasibleNoResult, # TODO improve
+    :mip_optimal        => MOI.Success,
+    :mip_unbounded      => MOI.UnboundedNoResult # TODO improve
+)
+
+
+
+## Result status
+
+# struct PrimalStatus <: AbstractSolverInstanceAttribute
+function MOI.cangetattribute(m::XpressSolverInstance, primal::MOI.PrimalStatus)
+    if primal.N != 1
+        return false
+    end
+    return true
+end
+function MOI.getattribute(m::XpressSolverInstance, primal::MOI.PrimalStatus)
+    if !is_mip(m.inner)
+        return MOI.FeasiblePoint
+    else
+        return MOI.FeasiblePoint
+    end
+end
+
+function MOI.cangetattribute(m::XpressSolverInstance, primal::MOI.DualStatus)
+    if primal.N != 1
+        return false
+    end
+    return true
+end
+function MOI.getattribute(m::XpressSolverInstance, primal::MOI.DualStatus)
+    if !is_mip(m.inner)
+        return MOI.FeasiblePoint
+    else
+        return MOI.FeasiblePoint
+    end
+end
```

Comment:
ditto. I realised that it's not actually that easy to map every CPLEX status to a termination/primal/dual triplet...